kong/tools/request_aware_table.lua

```diff
@@ -0,0 +1,89 @@
+local table_clear = require "table.clear"
+
+local get_phase = ngx.get_phase
+local fmt = string.format
+local ngx_var = ngx.var
+
+local NGX_VAR_PHASES = {
+  set = true,
+  rewrite = true,
+  balancer = true,
+  access = true,
+  content = true,
+  header_filter = true,
+  body_filter = true,
+  log = true,
+}
+
+
+--- Request aware table constructor
+-- Wraps an existing table (or creates a new one) with request-aware access
+-- logic to protect the underlying data from race conditions.
+-- @param data_table (optional) The target table to use as underlying data
+-- @param request_awareness_mode (optional) The mode for request awareness
+--        - "off": No request awareness mode
+--        - any other value, or nil: Control access for every r/w operation
+-- @return The newly created table with request-aware access
+local function new(data_table, request_awareness_mode)
+  if data_table and type(data_table) ~= "table" then
+    error("data_table must be a table", 2)
+  end
+
+  local allowed_request_id
+  local proxy = {}
+  local data  = data_table or {}
+
+  -- Check if access is allowed based on the request ID
+  local function enforce_sequential_access()
+    local curr_phase = get_phase()
+    if not NGX_VAR_PHASES[curr_phase] then
+      error(fmt("cannot enforce sequential access in %s phase", curr_phase), 2)
+    end
+
+    local curr_request_id = ngx_var.request_id
```

Comment:
We can use the `get_request` method, which is cheaper than using the `request_id`.

See: https://github.com/Kong/kong/blob/master/kong/resty/ctx.lua#L51

Reply:
Here we need to identify whether the table was accessed via different (multiple) requests, I don't think we can compare the output of `get_request()` (which should be a cdata pointing to a [ngx-http-request-t](https://www.nginx.com/resources/wiki/extending/api/http/#ngx-http-request-t) struct) to tell requests apart, or can we?

Reply:
You should be able to as cdata are essentially pointer addresses and should be comparable. Could you give it a test?